### DISCUSSION TITLE
how to use streaming response with anyio stream queue properly

### DISCUSSION THREAD

**Discussion by DeoLeung** [fastapi/fastapi#12061] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
send_stream, recv_stream = create_memory_object_stream[bytes]()
tg = create_task_group()
await tg.__aenter__()
tg.start_soon(generate_messages, send_stream)

async def response_stream():
  async with recv_stream:
    async for message in recv_stream:
      yield message

return StreamingResponse(response_stream(), background=BackgroundTask(tg.__aexit__))
```

### Description

I used to above code to generate the streaming response, my question is 

is there a simpler way to fire the `generate_messages` and leave it running?
or is there a simpler way to manage the task group?

previously I just using the `generate_messages` as generator and put it into `StreamingResponse`, but exceptions in AsyncGenerator has many problems, such as `async generator ignored GeneratorExit`

so I'm thinking the use a simpler queue method as the generator, hopefully to reduce the error and ease the maintenace.

### Operating System

macOS

### FastAPI Version

0.112.1

### Pydantic Version

2.8.2

### Python Version

Python 3.11.9

**Comment by YuriiMotov** [comment, marked as answer]
I think you don't need that construction. You can just put `generate_messages` into `StreamingResponse`.
The reason why you had `async generator ignored GeneratorExit` error was likely that you were catching all the exceptions (`except Exception:`).

To avoid such problems, you should catch specific exceptions instead of all exceptions.
Or, you can start with adding except block to re-raise `GeneratorExit`:
```
try:
    ...

except GeneratorExit as e: # Add these two lines
    raise e

except Exception as e:
    ...
```
